packages/core/src/domain/session/sessionCookieStore.ts

```diff
@@ -1,27 +1,76 @@
 import { CookieOptions, getCookie, setCookie } from '../../browser/cookie'
 import * as utils from '../../tools/utils'
+import { isExperimentalFeatureEnabled } from '../configuration/experimentalFeatures'
+import { monitor } from '../internalMonitoring/internalMonitoring'
 import { SessionState, SESSION_EXPIRATION_DELAY } from './sessionStore'
 
 const SESSION_ENTRY_REGEXP = /^([a-z]+)=([a-z0-9-]+)$/
 const SESSION_ENTRY_SEPARATOR = '&'
 
 export const SESSION_COOKIE_NAME = '_dd_s'
+const RETRY_DELAY = 1
 
-export function withCookieLockAccess({
-  process,
-  after,
-  options
-}: {
-  process: (cookieSession: SessionState) => SessionState | undefined
-  after?: (cookieSession: SessionState) => void,
+type Operations = {
   options: CookieOptions
-}) {
-  const initialSession = retrieveSession()
-  const processedSession = process(initialSession)
+  process: (cookieSession: SessionState) => SessionState | undefined
+  after?: (cookieSession: SessionState) => void
+}
+
+export function withCookieLockAccess(operations: Operations) {
+  let currentLock: string
+  let currentSession = retrieveSession()
+  if (isExperimentalFeatureEnabled('cookie-lock')) {
+    // if someone has lock, postpone
+    if (currentSession.lock) {
+      postpone(operations)
+      return
+    }
+    // acquire lock
+    currentLock = utils.generateUUID()
+    currentSession.lock = currentLock
+    setSession(currentSession, operations.options)
+    // if lock is not acquired, postpone
+    currentSession = retrieveSession()
+    if (currentSession.lock !== currentLock) {
+      postpone(operations)
+      return
+    }
+  }
+  let processedSession = operations.process(currentSession)
+  if (isExperimentalFeatureEnabled('cookie-lock')) {
+    // if lock corrupted after process, postpone
+    currentSession = retrieveSession()
+    if (currentSession.lock !== currentLock!) {
+      postpone(operations)
+      return
+    }
+  }
   if (processedSession) {
-    persistSession(processedSession, options)
+    persistSession(processedSession, operations.options)
+  }
+  if (isExperimentalFeatureEnabled('cookie-lock')) {
+    if (!processedSession || !utils.isEmptyObject(processedSession)) {
```

Comment:
<!-- Questions are appropriate if you have a potential concern but are not quite sure if it's relevant or not. Asking the author for clarification or investigation can lead to a quick resolution.  -->
**❓ question:** ‏ I wonder why there is a different behavior when writing the `processedSession` if it is empty or not. FMU, when the `processedSession` is not an empty object, we "persist" it with the `lock` attribute present, and then we remove the `lock` attribute, for integrity check. Don't we need the same integrity check when removing the session / writing an empty object?

Reply:
Indeed, the issue with writing an empty session and keeping the lock is that the session is not empty anymore since we need to keep track of the lock in the session.
We could handle that by tweaking the clear session logic or by requiring withCookieLockAccess clients to handle the lock but it would increase the complexity.
This tradeoff seems simpler and I'd expect less concurrency issues around session expiration.

If that sounds ok to you, I could add some comments to clarify that though.

Reply:
We could simply make sure the lock is present in the `processedSession` object by doing something like this in `withCookieLockAccess`: 

```ts
processedSession.lock = lock
```

Reply:
This approach is the one that I had in mind behind the "tweaking the clear session logic".
FMU, by implementing it as is, when wanting to clear the session, `persistSession` will put `lock=x&expire=${now + expirationDelay}` in the cookie.
So this session could be considered as active, which even if does not seem to be an issue now is not ideal.
I'd then be in favor of handling this case differently in order to not cause unexpected issue.
Then, to release the lock for this case, we would need to handle it differently to also set the cookie expiration delay at 0.

Let's discuss it synchronously if you want.